lib/puppet/resource.rb

```diff
@@ -323,6 +323,10 @@ def initialize(type, title = nil, attributes = EMPTY_HASH)
         send(attr.to_s + "=", value)
       end
 
+      if environment.is_a?(Puppet::Node::Environment) && environment != Puppet::Node::Environment::NONE
```

Comment:
Pretty sure the code paths will always set this attribute to a Puppet::Node::Environment but our tests will throw it in as a string if they are not testing environment handling.

This also could be a `NONE` environment since that will have the externalize_path method defined, but we never care to externalize paths for in that case, so I made it explicit.